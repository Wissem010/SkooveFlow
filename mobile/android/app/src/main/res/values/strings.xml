<resources>
    <string name="app_name">SkooveFlow</string>
</resources>
