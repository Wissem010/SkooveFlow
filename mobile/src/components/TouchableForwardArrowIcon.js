import {View, TouchableOpacity, Image} from 'react-native';
import React from 'react';

export default function TouchableForwardArrowIcon({style, onPress}) {
  return (
    <View style={style}>
      <TouchableOpacity onPress={onPress}>
        <Image source={require('../assets/arrow_forward.png')} />
      </TouchableOpacity>
    </View>
  );
}
