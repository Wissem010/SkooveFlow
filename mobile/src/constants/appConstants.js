export const radioButtonsData = [
  {id: 1, label: 'option 1', value: 'option1'},
  {id: 2, label: 'option 2', value: 'option2'},
  {id: 3, label: 'option 3', value: 'option3'},
  {id: 4, label: 'option 4', value: 'option4'},
  {id: 5, label: 'option 5', value: 'option5'},
];
