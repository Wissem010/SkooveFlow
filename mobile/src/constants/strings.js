const SCREEN_A = 'ScreenA';
const SCREEN_B = 'ScreenB';
const SCREEN_D = 'ScreenD';
const RETRY_MESSAGE = 'Try resubmitting your answer...';
const REDIRECT_MESSAGE = "You'll be redirected shortly...";
const RADIO_GROUP_SELECTION_ERR_MSG = 'Please selection an item from the list';

module.exports = {
  SCREEN_A: SCREEN_A,
  SCREEN_B: SCREEN_B,
  SCREEN_D: SCREEN_D,
  RETRY_MESSAGE: RETRY_MESSAGE,
  REDIRECT_MESSAGE: REDIRECT_MESSAGE,
  RADIO_GROUP_SELECTION_ERR_MSG: RADIO_GROUP_SELECTION_ERR_MSG,
};
