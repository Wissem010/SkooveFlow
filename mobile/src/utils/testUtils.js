import React from 'react';
import {render} from '@testing-library/react';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
import {expirementReducer} from '../redux/slices/expirementSlice';
import {selectionReducer} from '../redux/slices/selectionSlice';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: {
        expirementReducer: expirementReducer,
        selectionReducer: selectionReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {},
) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>;
  }

  return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
