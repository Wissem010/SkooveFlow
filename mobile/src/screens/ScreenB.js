//React imports
import {View, Text, StyleSheet} from 'react-native';
import React, {useEffect, useState} from 'react';
//Redux imports
import {useDispatch, useSelector} from 'react-redux';
import {radioButtonsData} from '../constants/appConstants';
import {
  resetScreenBState,
  rSubmitSelection,
} from '../redux/slices/selectionSlice';
//UI imports
import RadioGroup from 'react-native-radio-buttons-group';
import TouchableForwardArrowIcon from '../components/TouchableForwardArrowIcon';
import Spinner from '../components/Spinner';
//Constants imports
import {RADIO_GROUP_SELECTION_ERR_MSG, SCREEN_D} from '../constants/strings';

export default function ScreenB({navigation}) {
  const [radioButtons, setRadioButtons] = useState(radioButtonsData);
  const [selectionErrMessage, setSelectionErrMessage] = useState('');

  function onPressRadioButton(radioButtonsArray) {
    setRadioButtons(radioButtonsArray);
  }
  const dispatch = useDispatch();
  const {loading, message, err, nextScreen} = useSelector(
    state => state.selectionReducer,
  );

  useEffect(() => {
    return navigation.addListener('focus', () => {
      //
    });
  }, [navigation]);

  useEffect(() => {
    if (nextScreen == SCREEN_D) {
      navigation.navigate(SCREEN_D);
      dispatch(resetScreenBState());
    }
  }, [nextScreen]);

  const onPressArrow = () => {
    if (radioButtons.find(r => r.selected)) {
      dispatch(rSubmitSelection());
      setSelectionErrMessage('');
    } else {
      setSelectionErrMessage(RADIO_GROUP_SELECTION_ERR_MSG);
    }
  };

  return loading ? (
    <View style={styles.backgroundStyle}>
      <Spinner />
    </View>
  ) : (
    <View style={styles.backgroundStyle}>
      <View>
        {err && <Text style={styles.textStyle}>{message}</Text>}
        {selectionErrMessage && <Text>{selectionErrMessage}</Text>}
      </View>
      <RadioGroup radioButtons={radioButtons} onPress={onPressRadioButton} />
      <TouchableForwardArrowIcon
        style={styles.arrowStyle}
        onPress={onPressArrow}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  backgroundStyle: {
    backgroundColor: 'yellow',
    display: 'flex',
    alignContent: 'space-between',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: '100%',
  },
  arrowStyle: {
    alignSelf: 'flex-end',
    padding: 20,
  },
  textStyle: {
    textAlign: 'center',
    color: 'red',
    padding: 20,
    fontSize: 28,
  },
});
