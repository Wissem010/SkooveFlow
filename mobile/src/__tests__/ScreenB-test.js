import React from 'react';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';
import {renderWithProviders} from '../utils/testUtils';
import {
  resetScreenBState,
  rSubmitSelection,
  selectionReducer,
} from '../redux/slices/selectionSlice';
import ScreenB from '../screens/ScreenB';
import TouchableForwardArrowIcon from '../components/TouchableForwardArrowIcon';
import {RETRY_MESSAGE, SCREEN_D} from '../constants/strings';

//Mocking navigation functions
const mockNavigation = {
  navigate: jest.fn(),
  addListener: jest.fn(),
};

//Mocking redux functions
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const mockStore = {
  selectionReducer,
};
const useSelectorMock = reactRedux.useSelector;
const useDispatchMock = reactRedux.useDispatch;
//Initializing mocks
beforeEach(() => {
  useDispatchMock.mockImplementation(() => () => {});
  useSelectorMock.mockImplementation(selector => selector(mockStore));
});
afterEach(() => {
  useDispatchMock.mockClear();
  useSelectorMock.mockClear();
});

test('renders correctly', () => {
  const tree = renderer.create(<TouchableForwardArrowIcon />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should return the selectionReducer initial state', () => {
  expect(selectionReducer(undefined, {type: undefined})).toEqual({
    loading: false,
    message: '',
    err: false,
    nextScreen: '',
  });
});

test('should return the wanted state after dispatch resetScreenB action', () => {
  expect(selectionReducer(undefined, resetScreenBState)).toEqual({
    loading: false,
    message: '',
    err: false,
    nextScreen: '',
  });
});

describe('SELECTION SLICE TESTS', () => {
  it('should set loading true while action is pending', () => {
    const action = {type: rSubmitSelection.pending};
    const state = selectionReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: true,
      message: '',
      err: false,
      nextScreen: '',
    });
  });

  it('should set correct screen when action is fulfilled', () => {
    const action = {
      type: rSubmitSelection.fulfilled,
    };
    const state = selectionReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: false,
      message: '',
      err: false,
      nextScreen: SCREEN_D,
    });
  });

  it('should set error true when action is rejected', () => {
    const action = {type: rSubmitSelection.rejected};
    const state = selectionReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: false,
      message: RETRY_MESSAGE,
      err: true,
      nextScreen: '',
    });
  });
});
test('render screenB with providers correctly', () => {
  const tree = renderWithProviders(
    <ScreenB navigation={mockNavigation} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});


