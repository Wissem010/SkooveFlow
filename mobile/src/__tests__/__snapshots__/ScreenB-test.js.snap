// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render screenB with providers correctly 1`] = `
<View
  style={
    Object {
      "alignContent": "space-between",
      "alignItems": "center",
      "backgroundColor": "yellow",
      "display": "flex",
      "height": "100%",
      "justifyContent": "space-between",
    }
  }
>
  <View />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "column",
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "flexDirection": "row",
          },
          Object {
            "opacity": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "borderColor": "#444",
              "borderRadius": 12,
              "borderWidth": 2.5,
              "height": 24,
              "width": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontSize": 22,
            },
          ]
        }
      >
        option 1
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "flexDirection": "row",
          },
          Object {
            "opacity": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "borderColor": "#444",
              "borderRadius": 12,
              "borderWidth": 2.5,
              "height": 24,
              "width": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontSize": 22,
            },
          ]
        }
      >
        option 2
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "flexDirection": "row",
          },
          Object {
            "opacity": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "borderColor": "#444",
              "borderRadius": 12,
              "borderWidth": 2.5,
              "height": 24,
              "width": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontSize": 22,
            },
          ]
        }
      >
        option 3
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "flexDirection": "row",
          },
          Object {
            "opacity": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "borderColor": "#444",
              "borderRadius": 12,
              "borderWidth": 2.5,
              "height": 24,
              "width": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontSize": 22,
            },
          ]
        }
      >
        option 4
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "flexDirection": "row",
          },
          Object {
            "opacity": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "borderColor": "#444",
              "borderRadius": 12,
              "borderWidth": 2.5,
              "height": 24,
              "width": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontSize": 22,
            },
          ]
        }
      >
        option 5
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "padding": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/arrow_forward.png",
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`renders correctly 1`] = `
<View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/arrow_forward.png",
        }
      }
    />
  </View>
</View>
`;
